unified_planning/io/anml_reader.py

```
@@ -0,0 +1,810 @@
+            initial_default,
+        )
+
+    def _parse_objects(
+        self, objects_res: ParseResults, types_map: Dict[str, "up.model.Type"]
+    ) -> List["up.model.Object"]:
+        objects_type = self._parse_type_reference(objects_res["type"], types_map)
+        up_objects: List["up.model.Object"] = []
+        for name in objects_res["names"]:
+            assert isinstance(name, str), "parsing error"
+            up_objects.append(up.model.Object(name, objects_type))
+        return up_objects
+
+    def _parse_action(
+        self, action_res: ParseResults, types_map: Dict[str, "up.model.Type"]
+    ) -> "up.model.Action":
+        name = action_res["name"]
+        assert isinstance(name, str), "parsing error"
+        params = self._parse_parameters_def(action_res["parameters"], types_map)
+        action = up.model.DurativeAction(name, _parameters=params)
+        action_parameters: Dict[str, "up.model.Parameter"] = {
+            n: action.parameter(n) for n in params
+        }
+        self._populate_parsed_action_body(
+            action, action_res["body"], action_parameters, types_map
+        )
+        return action
+
+    def _populate_parsed_action_body(
+        self,
+        action: "up.model.DurativeAction",
+        action_body_res: ParseResults,
+        action_parameters: Dict[str, "up.model.Parameter"],
+        types_map: Dict[str, "Type"],
+    ) -> None:
+        for interval_and_exp in action_body_res:
+            relevant_words = {TK_DURATION, TK_ASSIGN, TK_INCREASE, TK_DECREASE, TK_WHEN}
+            contained_words = count_instances(interval_and_exp, relevant_words)
+            c_duration = contained_words[TK_DURATION]
+            c_when = contained_words[TK_WHEN]
+            c_assign = contained_words[TK_ASSIGN]
+            c_increase = contained_words[TK_INCREASE]
+            c_decrease = contained_words[TK_DECREASE]
+
+            if c_duration > 0:  # handle duration
```

If a duration expression is not in a form that is supported by the parser we should raise an exception with a clear error message